mempool/src/pool/reorg.rs

```diff
@@ -0,0 +1,162 @@
+}
+
+fn find_disconnected_txs(
+    chainstate: &dyn ChainstateInterface,
+    old_tip_id: Id<GenBlock>,
+    new_tip_id: Id<Block>,
+) -> Result<Vec<SignedTransaction>, ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    let old_tip_id = match old_tip_id.classify(chain_config) {
+        common::chain::GenBlockId::Genesis(_) => return Ok(Vec::new()),
+        common::chain::GenBlockId::Block(id) => id,
+    };
+
+    let common_id = {
+        let old_index = chainstate.get_block_index(&old_tip_id)?.ok_or(ReorgError::OldTipIndex)?;
+        let new_index = chainstate.get_block_index(&new_tip_id)?.ok_or(ReorgError::NewTipIndex)?;
+        let common_index = chainstate.last_common_ancestor(&old_index.into(), &new_index.into())?;
+        common_index.block_id()
+    };
+
+    // Short circuit the expensive processing below if no blocks are being disconnected
+    if old_tip_id == common_id {
+        return Ok(Vec::new());
+    }
+
+    // Collect IDs of trancsactions included in the newly connected chain
+    let mut connected_txs = BTreeSet::new();
+    for_each_block(chainstate, new_tip_id, common_id, |block| {
+        connected_txs.extend(block.transactions().iter().map(|t| t.transaction().get_id()))
+    })?;
+
+    // Collect txns that have been removed from the blockchain and not added on the new fork
+    let mut disconnected_txs = Vec::new();
+    for_each_block(chainstate, old_tip_id, common_id, |block| {
+        // We iterate blocks in the reverse order, so for consistent transaction input/output
+        // dependencies, take transacitons in reverse too,
+        let txns = block.transactions().iter().rev();
+        // We disregerd transactions that are re-added on the newly connected chain
+        let txns = txns.filter(|txn| !connected_txs.contains(&txn.transaction().get_id()));
+        disconnected_txs.extend(txns.cloned())
+    })?;
+
+    Ok(disconnected_txs)
+}
+
+fn fetch_disconnected_txs<M>(
+    mempool: &Mempool<M>,
+    new_tip: Id<Block>,
+) -> Result<Vec<SignedTransaction>, ReorgError> {
+    let chainstate = mempool.blocking_chainstate_handle();
+
+    let old_tip = mempool.tx_verifier.get_best_block_for_utxos().map_err(|_| ReorgError::OldTip)?;
+    if old_tip == new_tip {
+        return Ok(Vec::new());
+    }
+
+    chainstate.call(move |c| find_disconnected_txs(c.as_ref(), old_tip, new_tip))?
+}
+
+pub fn handle_new_tip<M: GetMemoryUsage>(mempool: &mut Mempool<M>, new_tip: Id<Block>) {
+    mempool.rolling_fee_rate.get_mut().set_block_since_last_rolling_fee_bump(true);
+
+    let mut disconnected_txs = fetch_disconnected_txs(mempool, new_tip)
+        .log_err_pfx("Fetching disconnected transactions on a reorg")
+        .unwrap_or_default();
+
+    // Take all the mempool previous transactions
+    let old_store = mem::replace(&mut mempool.store, MempoolStore::new());
+
+    // Discard the old tx verifier and replace it with a fresh one
+    mempool.tx_verifier = tx_verifier::create(
```

Comment:
I would create a method to do this instead of manually reassigning the new verifier in the interest of extensibility in the future.

Reply:
```
fn mempool.reset_tx_verifier() {
    ...
}
```